File: schemas/auth-rbac/convex/organization_invitations.ts
// organization_invitations: Pending invitations to join an organization.
// Separate from organization_members because an invitee may not have an account yet.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const organizationInvitations = defineTable({
  organizationId: v.id("organizations"),
  email: v.string(), // Invitee's email. They may or may not have an account yet.
  roleId: v.id("roles"), // The role they'll get upon acceptance.

  status: v.union(
    v.literal("pending"),
    v.literal("accepted"),
    v.literal("expired"),
    v.literal("revoked"),
  ),

  tokenHash: v.string(), // Hashed invitation token. Raw token sent in the invite email.
  inviterId: v.optional(v.id("users")), // Who sent the invitation. Null if system-generated.
  expiresAt: v.number(), // Typically 7 days.
  acceptedAt: v.optional(v.number()),
})
  .index("by_org_status", ["organizationId", "status"])
  .index("by_email", ["email"])
  .index("by_token_hash", ["tokenHash"]);

File: schemas/auth-rbac/spacetimedb/organization_invitations.rs
// organization_invitations: Pending invitations to join an organization.
// Separate from organization_members because an invitee may not have an account yet.
// See README.md for full design rationale.

use spacetimedb::{SpacetimeType, Timestamp};

#[derive(SpacetimeType, Clone)]
pub enum InvitationStatus {
    Pending,
    Accepted,
    Expired,
    Revoked, // An admin cancelled it before the invitee accepted.
}

#[spacetimedb::table(name = organization_invitations, public)]
pub struct OrganizationInvitation {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub organization_id: String, // FK → organizations.id (cascade delete)

    #[index(btree)]
    pub email: String, // Invitee's email. They may or may not have an account yet.

    pub role_id: String, // FK → roles.id (restrict delete). Role granted upon acceptance.

    pub status: InvitationStatus, // index(organization_id, status) for pending invitation listing.

    #[unique]
    pub token_hash: String, // Hashed invitation token. The raw token is sent in the invite email.

    pub inviter_id: Option<String>, // FK → users.id (set null on delete). Null if system-generated.

    pub expires_at: Timestamp, // Typically 7 days. After this, a new invitation must be requested.

    pub accepted_at: Option<Timestamp>, // When the invitee accepted.

    pub created_at: Timestamp,
}
